napi/src/threadsafe_function.rs

```diff
@@ -0,0 +1,157 @@
+use crate::{check_status, ptr, sys, Env, Function, Result, Value};
+use std::os::raw::{c_char, c_void};
+
+use sys::napi_threadsafe_function_call_mode;
+use sys::napi_threadsafe_function_release_mode;
+
+pub trait ToJs: Copy + Clone {
+  type Output;
+  type JsValue;
+
+  fn resolve(
+    &self,
+    env: &mut Env,
+    output: &mut Self::Output,
```

Comment:
why not just `Self::Output`?

Reply:
We don't use the `output` anymore so we can leave the owner of `output` to the user and avoid copying the `output`, can't we?

Reply:
> so we can leave the owner of output to the user 

Yes, but if you want give the ownership to the user in rust, you must `move` it, but not give the `borrowed` one to the user.

In most case, if the user want consume the `&mut Output`, they must call `*Output` or `Output.to_owned()` or something else, it's actually a `Copy` happened.
And in our case, `tsfn` and `async_callback`, we put the `Output` in the `heap memory` (Box) rather than `stack memory`. So pass `Box::from_raw(data)` to `resolve` function actually won't cause the `Copy` happened. It just `Copy` the smart pointer (Box) to the `resolve` under the hood.
The only problem is, we should use `output: Box<Self::Output>` to align the semantics. But I don't think this is a big problem here.

Reply:
> so we can leave the owner of output to the user 

Yes, but if you want give the ownership to the user in rust, you must `move` it, but not give the `borrowed` one to the user.

In most case, if the user want consume the `&mut Output`, they must call `*Output` or `Output.to_owned()` or something else, it's actually a `Copy` happened.

If the use case of `tsfn` is really performance sensitive, the user can choose `Box` or `Cell` to represent the `Output` themselves.

Reply:
Change to `Self::Output`. 

I think we can add another api that won't `Copy`. My use case is create a large `Buffer` of Node.js from `u8` without copying data and vice versa.

Reply:
> My use case is creating a large Buffer of Node.js from u8 without copying data and vice versa.

Yes, it seems to be a common use cases. Maybe we can create something like `NodeBuffer`, which mange the memory manually under the hood and could `deref/deref_mut` to `&[u8]/&mut [u8]`, and easy transform to `JsArrayBuffer/JsBuffer`